### ISSUE TITLE
gradient descent and performance issues with Tf 2.1

### ISSUE THREAD

**Issue by lordfiftyfive** [tensorflow/probability#891]
I am running a tensorflow algorithm using a gaussian process layer which is based on the example provided in google colab. I recently upgraded the the google colab to tensorflow 2.0. Unfortunately the GPU accelerator no longer seems to work. 

This may be some kind of bug because one of the warning I got said this

Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, export AUTOGRAPH_VERBOSITY=10) and attach the full output.

Even worse I am getting the following warnings which seem to be impacting my performance further. Before the upgrade I could get near 0 loss with 270 epochs. Now it has stopped improving altogether after a few hundred epochs and gets permanently stuck at 34.  
These warning appear to suggest that the lack of a GPU accelerator may not be the only thing affecting the performance of my algorithm

`WARNING:tensorflow:AutoGraph could not transform <function Model.make_train_function.<locals>.train_function at 0x7f5181e65c80> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Bad argument number for Name: 4, expecting 3
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert

WARNING:tensorflow:AutoGraph could not transform <function Model.make_train_function.<locals>.train_function at 0x7f5181e65c80> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Bad argument number for Name: 4, expecting 3
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert

WARNING: AutoGraph could not transform <function Model.make_train_function.<locals>.train_function at 0x7f5181e65c80> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Bad argument number for Name: 4, expecting 3
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_6/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_6/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_6/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_6/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_6/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_6/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_6/observation_noise_variance:0'] when minimizing the loss.
I am using a gaussian process layer based on the example provided and upgraded all the code in the google colab I am running it on to tensorflow 2.0. Unfortunately my GPU accelerator no longer seems to work as it gives this error message ``

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_6/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:AutoGraph could not transform <function Model.make_test_function.<locals>.test_function at 0x7f51826cd950> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Bad argument number for Name: 4, expecting 3
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert

WARNING:tensorflow:AutoGraph could not transform <function Model.make_test_function.<locals>.test_function at 0x7f51826cd950> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Bad argument number for Name: 4, expecting 3
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert

WARNING: AutoGraph could not transform <function Model.make_test_function.<locals>.test_function at 0x7f51826cd950> and will run it as-is.
Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: Bad argument number for Name: 4, expecting 3
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert`

**Comment by junpenglao** [comment]
Seems to me the biggest problem is `WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_6/observation_noise_variance:0'] when minimizing the loss.` my guess is that after you created the Variable you did computation on it (for example, doing `sd_log = tf.Variable(...)` and then `sd = tf.math.exp(sd_log)`). The solution is to use `tfp.util.TransformedVariable`:

https://github.com/tensorflow/probability/blob/6313adea719e93bc99addd003bff609040d2e510/tensorflow_probability/python/distributions/variational_gaussian_process.py#L517-L528

**Comment by lordfiftyfive** [comment]
I rewrote the kernel to look like this

```
dtype = np.float64
amplitude = tfp.util.TransformedVariable(
      1., tfb.Softplus(), dtype=dtype, name='amplitude')
length_scale = tfp.util.TransformedVariable(
      1., tfb.Softplus(), dtype=dtype, name='length_scale')
kernel = tfk.ExponentiatedQuadratic(
      amplitude=amplitude,
      length_scale=length_scale)
observation_noise_variance = tfp.util.TransformedVariable(
      1., tfb.Softplus(), dtype=dtype, name='observation_noise_variance')
```
And I am still getting the following output on runtime 

``
WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_8/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_8/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_8/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_8/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_8/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_8/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_8/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process_8/observation_noise_variance:0'] when minimizing the loss.
``

**Comment by smazcw3** [comment]
I am also getting this warning. Did this issue got resolved?

**Comment by junpenglao** [comment]
@csuter any idea?

**Comment by lordfiftyfive** [comment]
The issue has not been resolved. Furthermore I believe I know the source of this error. The devs for tensorflow-probability did not make the variational gaussian process layer "tape safe". Because it is not gradient tape safe that is why this happens. The devs on a different issue promised to fix this so I am assuming that this will be fixed when the new tensorflow probability pushes tensorflow-probability 0.10

**Comment by csuter** [comment]
Can you provide some code that reproduces the issue? The VGP distribution was made tape safe in late December. The corresponding Layer may still have an issue, or there could be some subtlety in how variables are introduced to the model; getting this right with Keras is trickier than I'd like...

If you can share some code we'll be better able to help. Also, are you currently using tensorflow-probability==0.9 or tfp-nightly when you encounter this issue?

**Comment by lordfiftyfive** [comment]
The VGP _distribution_ may be tape safe. I am certainly not disputing that. However I suspect the  VGP _layer_ is not tape safe. I am using the latest stable build of tensorflow-probability which is the 0.9 version. 

I am getting the following error based on the following code

```
loss = lambda y, rv_y: rv_y.variational_loss(
    y, kl_weight=np.array(batch_size, x.dtype) / x.shape[0])
#tf.keras.optimizers.Adam(1e-4) tf.optimizers.Adam(learning_rate=0.011)
model.compile(optimizer=tf.keras.optimizers.Adam(0.011), loss=loss)#tf.optimizers.Adam(learning_rate=0.01)
model.fit(x, y,epochs=370, verbose=True,validation_split=0.2)
```
```
Epoch 1/370
WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process/observation_noise_variance:0'] when minimizing the loss.

WARNING:tensorflow:Gradients do not exist for variables ['variational_gaussian_process/observation_noise_variance:0'] when minimizing the loss.
```

**Comment by csuter** [comment]
Thanks @lordfiftyfive -- I'm definitely ready to believe we have a bug in one or more places, including but not limited to lack of tape safety in both the distribution and the layer.

Can you provide a full, runnable code sample that reproduces the error? A [Colab](http://colab.research.google.com) link would work, if that's convenient for you.

**Comment by lordfiftyfive** [comment]
Here is the code for my kernel 
```
x = x.astype(np.float64)#tf.dtypes.cast(x, tf.int32) #
#x = tf.cast(x, tf.float32)
#x = tensor_util.convert_nonref_to_tensor(x, dtype=x.dtype)

class RBFKernelFn(tf.keras.layers.Layer):
  def __init__(self, **kwargs):
    super(RBFKernelFn, self).__init__(**kwargs)
    dtype = kwargs.get('dtype', None)
    self.amplitude = tfp.util.TransformedVariable(
      1., tfb.Softplus(), dtype=dtype, name='amplitude')
    self.length_scale = tfp.util.TransformedVariable(
      1., tfb.Softplus(), dtype=dtype, name='length_scale')
    
    
  def call(self, x):
    # Never called -- this is just a layer so it can hold variables
    # in a way Keras understands.
    #print(dtype)
    return x

  @property
  def kernel(self):

    
    return tfk.ExponentiatedQuadratic(
      amplitude=self.amplitude,
      length_scale=self.length_scale)
    observation_noise_variance = tfp.util.TransformedVariable(
      1., tfb.Softplus(), dtype=dtype, name='observation_noise_variance')

dtype = np.float64
amplitude = tfp.util.TransformedVariable(
      1., tfb.Softplus(), dtype=dtype, name='amplitude')
length_scale = tfp.util.TransformedVariable(
      1., tfb.Softplus(), dtype=dtype, name='length_scale')
kernel = tfk.ExponentiatedQuadratic(
      amplitude=amplitude,
      length_scale=length_scale)
observation_noise_variance = tfp.util.TransformedVariable(
      1., tfb.Softplus(), dtype=dtype, name='observation_noise_variance')
```

and this is the code for the neural network itself

```
x_tst = x[189::]
x_range = 237
num_distributions_over_Functions = 1
tf.keras.backend.set_floatx('float64')
#kernel = Brownian #tfp.positive_semidefinite_kernels.ExponentiatedQuadratic#MaternOneHalf()

model = tf.keras.Sequential([
    tf.keras.Input(shape=(1,14), dtype=np.float64),
    tf.keras.layers.LSTM(25,kernel_initializer='ones',activation='tanh', dtype = x.dtype, use_bias=True),
    #tf.keras.layers.InputLayer(input_shape=(10),dtype=x.dtype),#put a 1 before the 9 later
    tf.keras.layers.Dense(50,kernel_initializer='ones', use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(75,kernel_initializer='ones', use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(100,kernel_initializer='ones', use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(125,kernel_initializer='ones', use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(150,kernel_initializer='ones',use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(175,kernel_initializer='ones',use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(200,kernel_initializer='ones',use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(225,kernel_initializer='ones',use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(250,kernel_initializer='ones',use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(225,kernel_initializer='ones',use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(200,kernel_initializer='ones',use_bias=False),
    #goal is to eventually replace the first dense layer with an LSTM layer
    #tf.keras.layers.LSTM
    #tf.keras.layers.TimeDistributed(Dense(vocabulary)))
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(150,kernel_initializer='ones',use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(125,kernel_initializer='ones', use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(100,kernel_initializer='ones',use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(75,kernel_initializer='ones', use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(50,kernel_initializer='ones',use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(25, kernel_initializer='ones',use_bias=False,),
    tfp.layers.VariationalGaussianProcess(
    num_inducing_points=num_inducing_points, kernel_provider=RBFKernelFn(dtype=x.dtype) , event_shape=(1,),
    inducing_index_points_initializer=tf.compat.v1.constant_initializer(
            np.linspace(0,x_range, num=1125,
                        dtype=x.dtype)[..., np.newaxis]), unconstrained_observation_noise_variance_initializer=(tf.compat.v1.constant_initializer(np.log(np.expm1(1.)).astype(x.dtype))),variational_inducing_observations_scale_initializer=(tf.compat.v1.constant_initializer(np.log(np.expm1(1.)).astype(np.float64))), mean_fn=None,
    jitter=1e-06, convert_to_tensor_fn=tfp.distributions.Distribution.sample)

  
    #in unconstrained thing replace astype with tf.dtype thing.  RBFKernelFn(dtype=x.dtype)  Brownian(dtype=x.dtype)#tf.initializers.constant(-10.0)
])
`````

**Comment by lordfiftyfive** [comment]
So has anyone found what is causing the error?

**Comment by JimAva** [comment]
I'm having the same issue.  Any update?  Thanks.

**Comment by lordfiftysix** [comment]
Unfortunately no update. I have posted this multiple times on both the tensorflow and tensorflow probability issues. They are not responding

**Comment by SiegeLordEx** [comment]
https://github.com/tensorflow/probability/commit/67548a6d80a33f3fa43a2d0b5a96174f4a1e0b91 should fix the warning, please verify.

For future reference, things would go smoother in the future if there's a self-contained reproduction code. The code in https://github.com/tensorflow/probability/issues/891#issuecomment-628192336 was insufficient, as it had too many missing pieces (e.g. what's `x`?). The litmus test is whether you can just copy paste the code into a python file and run python on it without modifications.

**Comment by JimAva** [comment]
@SiegeLordEx - tfp-nightly fixed my warnings.  Thank you so much!

**Comment by lordfiftyfive** [comment]
Yep that did the trick
